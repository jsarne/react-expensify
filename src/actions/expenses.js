import firedb from '../firebase/firebase';

export const addExpense = (expense) => ({
  type: 'ADD_EXPENSE',
  expense: expense
});
export const startAddExpense = (expenseData = {}) => {
  return (dispatch) => {
    const {
      description='', 
      note='', 
      amount=0, 
      createdAt=0
    } = expenseData;
    const exp = {description, note, amount, createdAt};

    return firedb.ref('expenses')
      .push(exp)
      .then((fireExp) => {
        dispatch(addExpense({id: fireExp.key, ...exp}));
      });
  };
};

export const removeExpense = (id) => ({
  type: 'REMOVE_EXPENSE',
  id
});
export const startRemoveExpense = (id) => {
  return (dispatch) => {
    return firedb.ref(`expenses/${id}`).remove().then(() => {
      dispatch(removeExpense(id));
    });
  }
};

export const editExpense = (id, edits) => ({
    type: 'EDIT_EXPENSE',
    id,
    edits
});
export const startEditExpense = (id, edits) => {
  return (dispatch) => {
    return firedb.ref(`expenses/${id}`).update(edits).then(() => {
      dispatch(editExpense(id, edits));
    });
  }
};

export const setExpenses = (expenses) => ({
  type: 'SET_EXPENSES',
  expenses
});
export const startSetExpenses = () => {
  return (dispatch) => {
    return firedb.ref('expenses')
      .once('value')
      .then((fireExp) => {
        const storeExp = [];
        fireExp.forEach((fe) => {
          storeExp.push({id: fe.key, ...fe.val()});
        });
        dispatch(setExpenses(storeExp));
      });
  }
};
