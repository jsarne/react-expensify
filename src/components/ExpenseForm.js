import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';

export default class ExpenseForm extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      description: props.expense ? props.expense.description : '',
      note: props.expense ? props.expense.note : '',
      amount: props.expense ? (props.expense.amount/100).toString() : '',
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      calendarFocused: false,
      error: '',
      buttonText: props.source === 'add' ? 'Add Expense' : 'Save Expense'
    };
  }

  onDescriptionChange = (e) => {
    const description = e.target.value;
    this.setState(() => ({description}));
  }

  onNoteChange = (e) => {
    const note = e.target.value;
    this.setState(() => ({note}));
  }

  onAmountChange = (e) => {
    const amount = e.target.value;
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => ({amount}));
    }
  }

  onDateChange = (createdAt) => {
    if (createdAt) {
      this.setState(() => ({createdAt}));
    }
  }
  // same as above
  // onDateChange = (createdAt) => {
  //   this.setState(() => {
  //     return {createdAt};
  //   });
  // }

  onSubmit = (e) => {
    e.preventDefault();
    if (!this.state.description || !this.state.amount) {
      this.setState(() => ({error: 'Please provide description and amount.'}));
    } else {
      this.setState(() => ({error: ''}));
      this.props.onSubmit({
        description: this.state.description,
        amount: parseFloat(this.state.amount) * 100,
        note: this.state.note,
        createdAt: this.state.createdAt.valueOf()
      })
    }
  }

  onCalendarFocusChange = ({focused}) => {
    this.setState(() => ({calendarFocused: focused}));
  }

  render() {
    return (
      <form className="form" onSubmit={this.onSubmit}>
        {this.state.error && <p className="form__error">{this.state.error}</p>}
        <input 
          className="text-input"
          type="text" 
          placeholder="Description" 
          autoFocus 
          value={this.state.description}
          onChange={this.onDescriptionChange} />
        <input 
          type="text" 
          className="text-input"
          placeholder="Amount" 
          value={this.state.amount}
          onChange={this.onAmountChange} />
        <SingleDatePicker 
          date={this.state.createdAt}
          onDateChange={this.onDateChange}
          focused={this.state.calendarFocused}
          onFocusChange={this.onCalendarFocusChange}
          numberOfMonths={1}
          isOutsideRange={() => false}  // no days are outside the range, so no need for a complex function
          displayFormat="YYYY-MM-DD"
        />
        <textarea 
          className="textarea"
          placeholder="Add a note for your expense (optional)"
          value={this.state.note}
          onChange={this.onNoteChange} >
        </textarea>
        <div>
          <button className="button">{this.state.buttonText}</button>
        </div>
      </form>
    )
  }
}