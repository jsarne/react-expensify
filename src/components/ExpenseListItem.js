import React from 'react';
import {Link} from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeral';

// pair with <ExpenseListItem key={exp.id} expense={exp} /> in ExpenseList
// const ExpenseListItem = (props) => (
//   <div>
//     <h3>{props.expense.description}</h3>
//     <p>{props.expense.amount} - {props.expense.createdAt}</p>
//   </div>
// );
export const ExpenseListItem = ({id, description, amount, createdAt}) => (
  <Link className="list-item" to={`/edit/${id}`}>
    <div>
      <h3 className="list-item__title">{description}</h3>
      <span className="list-item__subtitle">{moment(createdAt).format('YYYY-MM-DD')}</span>
    </div>
    <h3 className="list-item__data">{numeral(amount/100).format('$0,0.00')}</h3>
  </Link>
);

export default ExpenseListItem;