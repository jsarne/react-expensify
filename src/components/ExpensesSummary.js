import React from 'react';
import {connect} from 'react-redux';
import numeral from 'numeral';
import selectExpenses from '../selectors/expenses';
import getExpensesTotal from '../selectors/expenses-total';

export const ExpensesSummary = (props) => {
  const expenseWord = props.expenses.length === 1 ? 'expense' : 'expenses';
  const formattedTotal = numeral(getExpensesTotal(props.expenses)/100).format('$0,0.00');
  return (
    <div>
      {
        props.expenses && 
        props.expenses.length > 0 && 
        <h1>Viewing {props.expenses.length} {expenseWord} totalling {formattedTotal}</h1>
      }
    </div>
  );
}

const mapStateToProps = (state) => {
  return {
    expenses: selectExpenses(state.expenses, state.filters)
  };
};

export default connect(mapStateToProps)(ExpensesSummary);