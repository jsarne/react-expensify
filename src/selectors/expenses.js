import moment from 'moment';

export default (expenses, {text, sortBy, startDate, endDate}) => {
  return expenses.filter((exp) => {
    const createdAtMmoent = moment(exp.createdAt);
    const descTextFound = exp.description.toLowerCase().includes(text.toLowerCase());
    const startDateInRange = startDate ? startDate.isSameOrBefore(createdAtMmoent, 'day') : true;
    const endDateInRange = endDate ? endDate.isSameOrAfter(createdAtMmoent, 'day') : true

    return descTextFound && startDateInRange && endDateInRange;
  }).sort((a, b) => {
    if (sortBy === 'date') {
      return a.createdAt < b.createdAt ? 1 : -1;  // larger createdAt - i.e. more recent - first
    } else if (sortBy === 'amount') {
      return a.amount < b.amount ? 1 : -1; // larger amount first
    }
  });
};