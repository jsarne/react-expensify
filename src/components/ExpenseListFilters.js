import React from 'react';
import {connect} from 'react-redux';
import { setTextFilter, sortByAmount, sortByDate, setStartDate, setEndDate } from '../actions/filters';
import { DateRangePicker } from 'react-dates';

export class ExpenseListFilters extends React.Component {
  state = {
    calendarFocused: null
  };

  onSortChange = (sortKey) => {
    switch (sortKey) {
      case "date": 
        this.props.sortByDate();
        return;
      case "amount": 
        this.props.sortByAmount();
        return;
    }
  }

  onDatesChange = ({startDate, endDate}) => {
    this.props.setStartDate(startDate);
    this.props.setEndDate(endDate);
  }

  onFocusChange = (calendarFocused) => {
    this.setState(() => ({calendarFocused}));
  }

  onTextChange = (e) => {
    this.props.setTextFilter(e.target.value);
  }

  render() {
    return (
      <div className="content-container">
        <div className="input-group">
          <div className="input-group__item">
            <input 
              className="text-input" 
              placeholder="Search Expenses"
              type="text" 
              value={this.props.filters.text} onChange={this.onTextChange}/>
          </div>
          <div className="input-group__item">
            <select className="select" value={this.props.filters.sortBy} onChange={(e) => this.onSortChange(e.target.value)}>
              <option value="date">Date</option>
              <option value="amount">Amount</option>
            </select>
          </div>
          <div className="input-group__item">
            <DateRangePicker 
              startDate={this.props.filters.startDate}
              endDate={this.props.filters.endDate}
              onDatesChange={this.onDatesChange}
              focusedInput={this.state.calendarFocused}
              onFocusChange={this.onFocusChange}
              showClearDates={true}
              numberOfMonths={1}
              isOutsideRange={() => false}
            />
          </div>
        </div>
      </div>  
    )
  }
}

const mapStateToProps = (state) => ({filters: state.filters});

const mapDispatchToProps = (dispatch) => ({
  setTextFilter: (text) => dispatch(setTextFilter(text)),
  sortByDate: () => dispatch(sortByDate()),
  sortByAmount: () => dispatch(sortByAmount()),
  setStartDate: (date) => dispatch(setStartDate(date)),
  setEndDate: (date) => dispatch(setEndDate(date))
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);
