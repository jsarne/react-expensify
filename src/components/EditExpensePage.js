import React from 'react';
import {connect} from 'react-redux';
import ExpenseForm from './ExpenseForm';
import {startEditExpense, startRemoveExpense} from '../actions/expenses';

export class EditExpensePage extends React.Component {
  onSubmit = (expense) => {
    this.props.editExpense(this.props.match.params.id, expense);
    this.props.history.push('/');
  }

  onRemoveClick = () => {
    this.props.removeExpense(this.props.match.params.id);
    this.props.history.push('/');
  }

  render() {
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <h1 className="page-header__title">Edit Expense</h1>
          </div>
        </div>
        <div className="content-container">
          <ExpenseForm source="edit" expense={this.props.expense} onSubmit={this.onSubmit} />
          <button className="button button--secondary" onClick={this.onRemoveClick}>Remove Expense</button>
        </div>
        <div>
        </div>
      </div>
    );  
  }
}

const mapStateToProps = (state, props) => ({
  expense: state.expenses.find((exp) => exp.id === props.match.params.id)
});

const mapDispatchToProps = (dispatch) => ({
  editExpense: (id, expense) => dispatch(startEditExpense(id, expense)),
  removeExpense: (id) => dispatch(startRemoveExpense(id))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);