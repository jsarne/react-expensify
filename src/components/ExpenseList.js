import React from 'react';
import {connect} from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses';

// export unconnected version as well so we can provide test data from fixtures, not from redux store
export const ExpenseList = (props) => (
  <div>
    {props.expenses.length === 0 ? (
      <p>No expenses</p>
    ) : (
      props.expenses.map(exp => <ExpenseListItem key={exp.id} {...exp} />)
    )}
  </div>
);

const mapStateToProps = (state) => {
  return {
    expenses: selectExpenses(state.expenses, state.filters)
  };
};

export default connect(mapStateToProps)(ExpenseList);
