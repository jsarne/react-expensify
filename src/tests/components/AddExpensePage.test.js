/* eslint-disable no-undef */
import React from 'react';
import {shallow} from 'enzyme';
import {AddExpensePage} from '../../components/AddExpensePage';
import expenses from '../fixtures/expenses';

let addExpenseSpy, hist, wrapper;

beforeEach(() => {
  addExpenseSpy = jest.fn();
  hist = {push: jest.fn()};
  wrapper = shallow(<AddExpensePage addExpense={addExpenseSpy} history={hist}/>);
});

test('render AddExpense', () => {
  expect(wrapper).toMatchSnapshot();
});

test('handle onSubmit', () => {
  wrapper.find('ExpenseForm').prop('onSubmit')(expenses[1]);
  expect(hist.push).toHaveBeenLastCalledWith('/');
  expect(addExpenseSpy).toHaveBeenLastCalledWith(expenses[1]);
});