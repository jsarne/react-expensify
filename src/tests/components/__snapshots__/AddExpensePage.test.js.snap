// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render AddExpense 1`] = `
<div>
  <div
    className="page-header"
  >
    <div
      className="content-container"
    >
      <h1
        className="page-header__title"
      >
        Add Expense
      </h1>
    </div>
  </div>
  <div
    className="content-container"
  >
    <ExpenseForm
      onSubmit={[Function]}
      source="add"
    />
  </div>
</div>
`;
